airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/S3StorageOperations.kt

```diff
@@ -460,6 +478,99 @@ open class S3StorageOperations(
         }
     }
 
+    override fun listExistingObjects(
+        namespace: String?,
+        streamName: String,
+        objectPath: String,
+        pathFormat: String,
+        currentGenerationId: Long
+    ): List<String> {
+        val regexFormat: Pattern =
+            Pattern.compile(getRegexFormat(namespace, streamName, pathFormat))
+        var keys = listOf<S3ObjectSummary>()
+        var objects: ObjectListing = listObjects(objectPath)
+        while (objects.objectSummaries.size > 0) {
+            keys =
+                keys +
+                    objects.objectSummaries.filter { obj: S3ObjectSummary ->
+                        regexFormat
+                            .matcher(
+                                obj.key,
+                            )
+                            .matches()
+                    }
+            if (objects.isTruncated) {
+                objects = s3Client.listNextBatchOfObjects(objects)
+            } else {
+                break
+            }
+        }
+        if (currentGenerationId != -1L) {
+            keys = sliceObjectsByMetadata(keys, currentGenerationId)
+        }
+
+        return keys.map { obj: S3ObjectSummary -> obj.key }
+    }
+
+    private fun sliceObjectsByMetadata(
+        keyList: List<S3ObjectSummary>,
+        generationId: Long
+    ): List<S3ObjectSummary> {
+        val bucket = s3Config.bucketName!!
+        // Sort the list by lastModified first
+        val sortedList =
+            keyList.sortedWith(
+                Comparator.comparingLong { o: S3ObjectSummary -> o.lastModified.time }
+            )
+        if (sortedList.isEmpty()) return listOf()
+
+        var low = 0
+        var high = sortedList.size - 1
+
+        // For comparison purposes we can assume that anything with null genId is to be provided in
+        // the sliced list
+        val lowGenId = getGenerationIdFromMetadata(bucket, sortedList[low].key).getOrDefault(-1L)
+        val highGenId = getGenerationIdFromMetadata(bucket, sortedList[high].key).getOrDefault(-1L)
+
+        // We have data to filter from previous generations only if highestGenId is equal to
+        // provided
+        // genId and lowGenId is less that the provided genId
+        if (highGenId == generationId) {
+            if (lowGenId < generationId) {
+                logger.info { "Filtering files from previous generation" }
+                // modified binary search to find the index where low crosses high and use the
+                // inclusive
+                // index as
+                // the point to slice
+                while (low <= high) {
+                    // right shift by 1 is midpoint between 2 integers
+                    val mid = (low + high).ushr(1)
```

Comment:
this is cool, but why not just `(low + high) / 2`? 😅

Reply:
[just for nerd fun](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/src/kotlin/collections/Collections.kt#L337).

Reply:
lol. fair enough